## Input & Premise
- lines of digits
- each line is a 'bank', I have to select **two** digits and that gives the bank it's joltage
  - if line is `123456`, and I select the `3` and `5`, then the bank would have 35 jolts
- need to find maximum output joltage. So `11811988` would have `98` as highest jolt selection
  - total joltage is sum of bank joltages

## Thoughts
- highest answer has highest digit first, then smaller digit.    i.e. 97 vs 79
- does it suffice to find highest 2 digits
  - almost, if highest digit is to the right of smaller digit, need next highest to the right
    - unless highest digit is last digit, then highest two work

In [14]:
infile = open("input.txt", "r")
# firstLine = infile.readline().strip()
# firstLine
# this approach has smaller always lagging behind higher, meaning it 
lineNum = 1
totalJoltage = 0
for line in infile:
    # print(f"working on line: {lineNum}\tlength: {len(line.strip())}")
    line = line.strip()
    # get top three highest values from a list
    biggest, bigger, big = 0,0,0
    biggestIdx, biggerIdx, bigIdx = -1,-1,-1
    for idx, val in enumerate(line):
        val = int(val)
        if val > biggest:
            biggest, biggestIdx = val, idx

    for idx, val in enumerate(line):
        val = int(val)
        if val > bigger and idx != biggestIdx:
            bigger, biggerIdx = val, idx

    for idx, val in enumerate(line):
        val = int(val) 
        if val > big and idx != biggestIdx and idx != biggerIdx:
            big, bigIdx = val, idx
    
    

    # print(f"biggest: {biggest} at {biggestIdx}, bigger: {bigger} at {biggerIdx}, big: {big} at {bigIdx}")

    # now select top two 
    bankJoltage = ""
    if biggestIdx > biggerIdx and bigIdx > biggestIdx:   # case of 6 7 6, 76 is better than 67
        # print(f"00  line {lineNum} selecting {biggest} at {biggestIdx} and {big} at {bigIdx}")
        bankJoltage = str(biggest) + str(big)
    elif biggestIdx < biggerIdx:            # case of 7 6 7, 77 is best
        # print(f"11  line {lineNum} selecting {biggest} at {biggestIdx} and {bigger} at {biggerIdx}")
        bankJoltage = str(biggest) +  str(bigger)
    elif biggestIdx == len(line)-1:         # case of 6 7 8, 78 is best
        # print(f"22  line {lineNum} selecting {bigger} at {biggerIdx} and {biggest} at {biggestIdx}")
        bankJoltage = str(bigger) + str(biggest)
    elif biggestIdx > biggerIdx and biggestIdx > bigIdx:  # case of 6 7 8 8, 88 is best but not registered in biggest,bigger,big so repeat big search from right of biggestIdx
        nextBiggest, nextBiggestIdx = 0, -1
        for idx in range(biggestIdx+1, len(line)):
            val = int(line[idx])
            if val > nextBiggest:
                nextBiggest, nextBiggestIdx = val, idx
        # print(f"33  line {lineNum} selecting {biggest} at {biggestIdx} and {nextBiggest} at {nextBiggestIdx}")
        bankJoltage = str(biggest) + str(nextBiggest)
    else:            
        print(f"XXX line {lineNum} did not fall into any case XXX")       

    totalJoltage += int(bankJoltage)

    lineNum += 1

print(f"Total bank joltage: {totalJoltage}")

Total bank joltage: 17376


## Part 2

- same premise but now I 'turn on' 12 batteries instead of 2
- need a different approach
- thinking along the lines of:
  - find highest digit atleast 12 spaces out from end
  - if highestIdx further than 11 out, search from there for next highest 10 out and repeat until last digit
  - for 1 2 3 4 5 8 6 5 2 3 1 2 6 and only 4 batteries that would make it
  - **8** 6 5 2 3 1 2 6
  - **8** **6** **5** **6**

In [ ]:
example = ["987654321111111","811111111111119","234234234234278","818181911112111"]
exampleAnswers = ["987654321111","811111111119","434234234278","888911112111"]
myExAns = []


infile = open("input.txt", "r")
lineNum = 1
totalJoltage = 0
for line in infile:
    bankJoltage = ""
    line = line.strip()
    lineLength = len(line) 
    highest = -1
    highestIdx = -1
    for i in range(12, 0, -1): # for each of the 12 digits; starting with first digit means we need at least 12 spaces out from end
       
        if highestIdx + i == lineLength:
            bankJoltage += line[highestIdx:]
            break

        # from right to left, find highest digit and keep newest (smallest index)
        for idx in range(lineLength-1, highestIdx, -1 ): 
            val = int(line[idx])
            if val >= highest and idx <= lineLength - i: # ensure enough space is left for remaining digits
                highest = val
                highestIdx = idx

        bankJoltage += str(highest)
        # print(f"i: {i}\thighest: {highest}\thighestIdx: {highestIdx}\tbankJoltage: {bankJoltage}")
        highest = -1

    # print(f"line {lineNum}\t\t\t\tbankJoltage: {bankJoltage}")
    lineNum += 1
    totalJoltage += int(bankJoltage)
    myExAns.append(bankJoltage)

print(f"Total bank joltage: {totalJoltage}")

i: 12	highest: 9	highestIdx: 16	bankJoltage: 9
i: 11	highest: 9	highestIdx: 75	bankJoltage: 99
i: 10	highest: 8	highestIdx: 76	bankJoltage: 998
i: 9	highest: 6	highestIdx: 81	bankJoltage: 9986
i: 8	highest: 4	highestIdx: 83	bankJoltage: 99864
i: 7	highest: 8	highestIdx: 93	bankJoltage: 998648
i: 6	highest: 5	highestIdx: 94	bankJoltage: 9986485
i: 5	highest: 9	highestIdx: 95	bankJoltage: 99864859
i: 4	highest: 3	highestIdx: 96	bankJoltage: 998648593
i: 3	highest: 8	highestIdx: 97	bankJoltage: 9986485938
i: 2	highest: 2	highestIdx: 98	bankJoltage: 99864859382
i: 1	highest: 4	highestIdx: 99	bankJoltage: 998648593824
line 1				bankJoltage: 998648593824
i: 12	highest: 9	highestIdx: 23	bankJoltage: 9
i: 11	highest: 8	highestIdx: 27	bankJoltage: 98
i: 10	highest: 8	highestIdx: 43	bankJoltage: 988
i: 9	highest: 8	highestIdx: 51	bankJoltage: 9888
i: 8	highest: 8	highestIdx: 67	bankJoltage: 98888
i: 7	highest: 8	highestIdx: 76	bankJoltage: 988888
i: 6	highest: 8	highestIdx: 79	bankJoltage: 98888

In [30]:
print(f"\t    example     vs       me        |     same")
print(f"first     {exampleAnswers[0]:15}  {myExAns[0]:15}      {exampleAnswers[0]==myExAns[0]} ")
print(f"second    {exampleAnswers[1]:15}  {myExAns[1]:15}      {exampleAnswers[1]==myExAns[1]} ")
print(f"third     {exampleAnswers[2]:15}  {myExAns[2]:15}      {exampleAnswers[2]==myExAns[2]} ")
print(f"fourth    {exampleAnswers[3]:15}  {myExAns[3]:15}      {exampleAnswers[3]==myExAns[3]} ")

	    example     vs       me        |     same
first     987654321111     998648593824         False 
second    811111111119     988888887786         False 
third     434234234278     777777777789         False 
fourth    888911112111     672222622222         False 


# Done!!

### Part 1
- Answer: 17376
- took a minute to figure out approach, ended up combining 'top three' with checking to right of highest

### Part 2
- Answer: 172119830406258
- Approach felt much more intuitive for a larger case than two 'batteries' actually